common/cuda_hip/matrix/bccoo_helper.hpp.inc

```diff
@@ -0,0 +1,263 @@
+        }
+    }
+    // segmented scan
+    if (tile_block.any(new_value)) {
+        bool is_first_in_segment = segment_scan<subwarp_size>(
+            tile_block, idxs.row, temp_val,
+            [](ValueType a, ValueType b) { return a + b; });
+        if (is_first_in_segment) {
+            atomic_add(&(c[idxs.row * c_stride + column_id]), scale(temp_val));
+        }
+        temp_val = zero<ValueType>();
+    }
+}
+
+
+template <typename IndexType, typename ValueType>
+inline GKO_ATTRIBUTES void get_block_position_value(
+    const size_type pos, const uint8* chunk_data, compr_blk_idxs& blk_idxs,
+    size_type& row, size_type& col, ValueType& val)
+{
+    row = blk_idxs.row_frs;
+    col = blk_idxs.col_frs;
+    if (blk_idxs.mul_row) {
+        if (blk_idxs.row_16bits) {
+            row += get_value_chunk<uint16>(
+                chunk_data, blk_idxs.shf_row + pos * sizeof(uint16));
+        } else {
+            row += get_value_chunk<uint8>(chunk_data, blk_idxs.shf_row + pos);
+        }
+    }
+    if (blk_idxs.col_8bits) {
+        col += get_value_chunk<uint8>(chunk_data, blk_idxs.shf_col + pos);
+    } else if (blk_idxs.col_16bits) {
+        col += get_value_chunk<uint16>(chunk_data,
+                                       blk_idxs.shf_col + pos * sizeof(uint16));
+    } else {
+        col += get_value_chunk<uint32>(chunk_data,
+                                       blk_idxs.shf_col + pos * sizeof(uint32));
+    }
+    val = get_value_chunk<ValueType>(
+        chunk_data, blk_idxs.shf_val + pos * sizeof(ValueType));
+}
+
+
+template <typename IndexType, typename ValueType, typename Closure>
+inline GKO_ATTRIBUTES void get_block_position_value_put(
+    const size_type pos, uint8* chunk_data, compr_blk_idxs& blk_idxs,
+    size_type& row, size_type& col, ValueType& val, Closure finalize_op)
+{
+    row = blk_idxs.row_frs;
+    col = blk_idxs.col_frs;
+    if (blk_idxs.mul_row) {
+        if (blk_idxs.row_16bits) {
+            row += get_value_chunk<uint16>(
+                chunk_data, blk_idxs.shf_row + pos * sizeof(uint16));
+        } else {
+            row += get_value_chunk<uint8>(chunk_data, blk_idxs.shf_row + pos);
+        }
+    }
+    if (blk_idxs.col_8bits) {
+        col += get_value_chunk<uint8>(chunk_data, blk_idxs.shf_col + pos);
+    } else if (blk_idxs.col_16bits) {
+        col += get_value_chunk<uint16>(chunk_data,
+                                       blk_idxs.shf_col + pos * sizeof(uint16));
+    } else {
+        col += get_value_chunk<uint32>(chunk_data,
+                                       blk_idxs.shf_col + pos * sizeof(uint32));
+    }
+    val = get_value_chunk<ValueType>(
+        chunk_data, blk_idxs.shf_val + pos * sizeof(ValueType));
+    auto new_val = finalize_op(val);
+    set_value_chunk<ValueType>(
+        chunk_data, blk_idxs.shf_val + pos * sizeof(ValueType), new_val);
+}
+
+
+template <typename IndexType, typename ValueType1, typename ValueType2,
+          typename Closure>
```

Comment:
I have changed different templates, to adapt them to the last changes:
- ValueType1 and ValueType2 are changed to ValueTypeSrc and ValueTypeRes in get_block_position_value_put.
- IndexType1 and IndexType2 are changed to IndexTypeRow and IndexTypeCol in loop_block_multi_row.
- I have removed IndexType in all get_block_position_value* methods, since it not used.
In the next step, I will remove get_block_position_value* methods, defining methods similar to loop_block_single_row and loop_block_multi_row to move them as far as possible

Please, resolve if you agree

Reply:
So, I will try to use the switch, avoiding the multi_if code

Reply:
Please, resolve if you agree